# Premier League Match Predictor

A test of various machine learning models at predicting the winners of premier league games

In [1]:
import pandas as pd

## Data Preprocessing

In [2]:
matches = pd.read_csv("matches.csv", index_col=0)

In [3]:
matches.head()

Date   Time            Comp        Round  Day Venue Result   GF   GA  \
0  2023-08-13  16:30  Premier League  Matchweek 1  Sun  Away      D  1.0  1.0   
1  2023-08-19  15:00  Premier League  Matchweek 2  Sat  Home      W  3.0  1.0   
2  2023-08-27  16:30  Premier League  Matchweek 3  Sun  Away      W  2.0  1.0   
3  2023-09-03  14:00  Premier League  Matchweek 4  Sun  Home      W  3.0  0.0   
4  2023-09-16  12:30  Premier League  Matchweek 5  Sat  Away      W  3.0  1.0   

        Opponent  ...  Blocks   Int  Tkl+Int   Clr SoTA Saves OppCrosses  \
0        Chelsea  ...    19.0  10.0       29  39.0  4.0   3.0       18.0   
1    Bournemouth  ...    13.0  11.0       26  15.0  5.0   4.0        6.0   
2  Newcastle Utd  ...    12.0   8.0       20  19.0  8.0   7.0       11.0   
3    Aston Villa  ...     5.0   8.0       22  11.0  3.0   3.0        9.0   
4         Wolves  ...     9.0   9.0       33  20.0  2.0   1.0       11.0   

  CrossesStp  Season       Team  
0        1.0    2023  Liverpool  
1        1.0    2023  Liverpool  
2        0.0    2023  Liverpool  
3        1.0    2023  Liverpool  
4        0.0    2023  Liverpool  

[5 rows x 35 columns]

In [4]:
matches.shape

(2056, 35)

In [5]:
matches["Team"].value_counts()

Team
Liverpool                   103
Manchester City             103
Arsenal                     103
Aston Villa                 103
Everton                     103
Manchester United           103
West Ham United             103
Newcastle United            103
Brighton and Hove Albion    103
Wolverhampton Wanderers     103
Brentford                   103
Crystal Palace              103
Chelsea                     102
Tottenham Hotspur           102
Leicester City               76
Southampton                  76
Leeds United                 76
Fulham                       65
Nottingham Forest            65
Burnley                      65
Bournemouth                  64
Watford                      38
Norwich City                 38
Sheffield United             27
Luton Town                   26
Name: count, dtype: int64

In [6]:
matches["Round"].value_counts()

Round
Matchweek 1     60
Matchweek 14    60
Matchweek 27    60
Matchweek 25    60
Matchweek 24    60
Matchweek 23    60
Matchweek 22    60
Matchweek 21    60
Matchweek 2     60
Matchweek 19    60
Matchweek 18    60
Matchweek 16    60
Matchweek 15    60
Matchweek 20    60
Matchweek 13    60
Matchweek 7     60
Matchweek 12    60
Matchweek 4     60
Matchweek 5     60
Matchweek 6     60
Matchweek 3     60
Matchweek 8     60
Matchweek 10    60
Matchweek 11    60
Matchweek 9     60
Matchweek 26    58
Matchweek 17    58
Matchweek 28    40
Matchweek 32    40
Matchweek 37    40
Matchweek 36    40
Matchweek 35    40
Matchweek 31    40
Matchweek 34    40
Matchweek 33    40
Matchweek 30    40
Matchweek 29    40
Matchweek 38    40
Name: count, dtype: int64

In [7]:
matches.dtypes

Date             object
Time             object
Comp             object
Round            object
Day              object
Venue            object
Result           object
GF              float64
GA              float64
Opponent         object
xG              float64
xGA             float64
Poss            float64
Attendance      float64
Captain          object
Formation        object
Referee          object
Match Report     object
Notes           float64
Sh              float64
SoT             float64
Dist            float64
FK              float64
PK                int64
PKatt             int64
Blocks          float64
Int             float64
Tkl+Int           int64
Clr             float64
SoTA            float64
Saves           float64
OppCrosses      float64
CrossesStp      float64
Season            int64
Team             object
dtype: object

In [8]:
matches["Date"] = pd.to_datetime(matches["Date"])

In [9]:
matches["Venue_Code"] = matches["Venue"].astype("category").cat.codes

In [10]:
matches.head()

Date   Time            Comp        Round  Day Venue Result   GF   GA  \
0 2023-08-13  16:30  Premier League  Matchweek 1  Sun  Away      D  1.0  1.0   
1 2023-08-19  15:00  Premier League  Matchweek 2  Sat  Home      W  3.0  1.0   
2 2023-08-27  16:30  Premier League  Matchweek 3  Sun  Away      W  2.0  1.0   
3 2023-09-03  14:00  Premier League  Matchweek 4  Sun  Home      W  3.0  0.0   
4 2023-09-16  12:30  Premier League  Matchweek 5  Sat  Away      W  3.0  1.0   

        Opponent  ...   Int  Tkl+Int   Clr  SoTA Saves OppCrosses CrossesStp  \
0        Chelsea  ...  10.0       29  39.0   4.0   3.0       18.0        1.0   
1    Bournemouth  ...  11.0       26  15.0   5.0   4.0        6.0        1.0   
2  Newcastle Utd  ...   8.0       20  19.0   8.0   7.0       11.0        0.0   
3    Aston Villa  ...   8.0       22  11.0   3.0   3.0        9.0        1.0   
4         Wolves  ...   9.0       33  20.0   2.0   1.0       11.0        0.0   

  Season       Team  Venue_Code  
0   2023  Liverpool           0  
1   2023  Liverpool           1  
2   2023  Liverpool           0  
3   2023  Liverpool           1  
4   2023  Liverpool           0  

[5 rows x 36 columns]

In [11]:
matches["Opp_Code"] = matches["Opponent"].astype("category").cat.codes

In [12]:
matches["Hour"] = matches["Time"].str.replace(":.+", "", regex=True).astype("int")

In [13]:
matches["Day_Code"] = matches["Date"].dt.dayofweek

In [14]:
matches.head()

Date   Time            Comp        Round  Day Venue Result   GF   GA  \
0 2023-08-13  16:30  Premier League  Matchweek 1  Sun  Away      D  1.0  1.0   
1 2023-08-19  15:00  Premier League  Matchweek 2  Sat  Home      W  3.0  1.0   
2 2023-08-27  16:30  Premier League  Matchweek 3  Sun  Away      W  2.0  1.0   
3 2023-09-03  14:00  Premier League  Matchweek 4  Sun  Home      W  3.0  0.0   
4 2023-09-16  12:30  Premier League  Matchweek 5  Sat  Away      W  3.0  1.0   

        Opponent  ...  SoTA  Saves  OppCrosses  CrossesStp Season       Team  \
0        Chelsea  ...   4.0    3.0        18.0         1.0   2023  Liverpool   
1    Bournemouth  ...   5.0    4.0         6.0         1.0   2023  Liverpool   
2  Newcastle Utd  ...   8.0    7.0        11.0         0.0   2023  Liverpool   
3    Aston Villa  ...   3.0    3.0         9.0         1.0   2023  Liverpool   
4         Wolves  ...   2.0    1.0        11.0         0.0   2023  Liverpool   

  Venue_Code Opp_Code  Hour  Day_Code  
0          0        6    16         6  
1          1        2    15         5  
2          0       16    16         6  
3          1        1    14         6  
4          0       24    12         5  

[5 rows x 39 columns]

In [15]:
matches["target"] = (matches["Result"]=="W").astype("int")

In [2]:
#If you wish to include draws as outcomes, use this line instead of above
#matches["target"] = matches["Result"].map({'W':2 , 'D':1, 'L':0})

In [17]:
matches["target"]

0     0
1     1
2     1
3     1
4     1
     ..
38    0
39    0
40    0
41    0
42    0
Name: target, Length: 2056, dtype: int32

In [18]:
matches.head()

Date   Time            Comp        Round  Day Venue Result   GF   GA  \
0 2023-08-13  16:30  Premier League  Matchweek 1  Sun  Away      D  1.0  1.0   
1 2023-08-19  15:00  Premier League  Matchweek 2  Sat  Home      W  3.0  1.0   
2 2023-08-27  16:30  Premier League  Matchweek 3  Sun  Away      W  2.0  1.0   
3 2023-09-03  14:00  Premier League  Matchweek 4  Sun  Home      W  3.0  0.0   
4 2023-09-16  12:30  Premier League  Matchweek 5  Sat  Away      W  3.0  1.0   

        Opponent  ...  Saves  OppCrosses  CrossesStp  Season       Team  \
0        Chelsea  ...    3.0        18.0         1.0    2023  Liverpool   
1    Bournemouth  ...    4.0         6.0         1.0    2023  Liverpool   
2  Newcastle Utd  ...    7.0        11.0         0.0    2023  Liverpool   
3    Aston Villa  ...    3.0         9.0         1.0    2023  Liverpool   
4         Wolves  ...    1.0        11.0         0.0    2023  Liverpool   

  Venue_Code Opp_Code Hour  Day_Code  target  
0          0        6   16         6       0  
1          1        2   15         5       1  
2          0       16   16         6       1  
3          1        1   14         6       1  
4          0       24   12         5       1  

[5 rows x 40 columns]

## Training the Random Forest

In [19]:
from sklearn.ensemble import RandomForestClassifier

In [20]:
rf = RandomForestClassifier(n_estimators=80, min_samples_split=12, random_state=1)

In [21]:
train = matches[matches["Date"] < '2023-09-01']

In [22]:
test = matches[matches["Date"] > '2023-09-01']

In [23]:
predictors = ["Venue_Code", "Opp_Code", "Hour", "Day_Code"]

In [24]:
rf.fit(train[predictors],train["target"])

RandomForestClassifier(min_samples_split=12, n_estimators=80, random_state=1)

In [25]:
preds = rf.predict(test[predictors])

In [26]:
from sklearn.metrics import accuracy_score

In [27]:
acc = accuracy_score(test["target"], preds)

In [28]:
acc

0.6029411764705882

In [29]:
combined = pd.DataFrame(dict(actual=test["target"], prediction = preds))

In [30]:
pd.crosstab(index=combined["actual"], columns=combined["prediction"])

prediction    0   1
actual             
0           223  63
1           126  64

## Improving Predictions with Rolling Averages of Match stats

In [31]:
grouped_matches = matches.groupby("Team")

In [32]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("Date")
    rolling_stats = group[cols].rolling(3, closed='left').mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset=new_cols)
    return group

In [33]:
cols = ["GF", "GA", "xG", "xGA", "Poss", "Sh", "SoT", "Dist", "FK", "PK", "PKatt", "Blocks", "Int", "Tkl+Int", "Clr", "SoTA", "Saves", "OppCrosses", "CrossesStp"]
new_cols = [f"{c}_rolling" for c in cols]

In [34]:
matches_rolling = matches.groupby("Team").apply(lambda x: rolling_averages(x, cols, new_cols), include_groups=True)

C:\Users\lukeg\AppData\Local\Temp\ipykernel_8308\3370473215.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  matches_rolling = matches.groupby("Team").apply(lambda x: rolling_averages(x, cols, new_cols), include_groups=True)


In [35]:
matches_rolling.head()

Date   Time            Comp        Round  Day Venue Result  \
Team                                                                         
Arsenal 4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Home      W   
        5 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Away      W   
        7 2021-09-26  16:30  Premier League  Matchweek 6  Sun  Home      W   
        8 2021-10-02  17:30  Premier League  Matchweek 7  Sat  Away      D   
        9 2021-10-18  20:00  Premier League  Matchweek 8  Mon  Home      D   

            GF   GA        Opponent  ...  PK_rolling  PKatt_rolling  \
Team                                 ...                              
Arsenal 4  1.0  0.0    Norwich City  ...         0.0            0.0   
        5  1.0  0.0         Burnley  ...         0.0            0.0   
        7  3.0  1.0       Tottenham  ...         0.0            0.0   
        8  0.0  0.0        Brighton  ...         0.0            0.0   
        9  2.0  2.0  Crystal Palace  ...         0.0            0.0   

           Blocks_rolling  Int_rolling Tkl+Int_rolling Clr_rolling  \
Team                                                                 
Arsenal 4       14.666667     8.333333       21.000000   19.000000   
        5       16.333333     7.000000       22.000000   20.000000   
        7       13.333333     8.666667       23.000000   29.000000   
        8       12.666667     8.000000       26.666667   25.666667   
        9       12.666667    10.666667       26.666667   26.333333   

          SoTA_rolling Saves_rolling  OppCrosses_rolling  CrossesStp_rolling  
Team                                                                          
Arsenal 4     6.000000      3.000000                16.0            1.666667  
        5     5.333333      3.000000                18.0            1.333333  
        7     4.666667      3.000000                24.0            2.000000  
        8     2.666667      2.333333                21.0            1.666667  
        9     3.000000      2.666667                23.0            2.000000  

[5 rows x 59 columns]

In [36]:
matches_rolling = matches_rolling.droplevel('Team')

In [37]:
matches_rolling.index = range(matches_rolling.shape[0])

In [38]:
def make_predictions(data, predictors):
    train = data[data["Date"] < '2023-10-01']
    test = data[data["Date"] > '2023-10-01']
    rf.fit(train[predictors],train["target"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test["target"], prediction = preds))
    accuracy_score(test["target"], preds)
    return combined, acc

In [39]:
new_combined, new_acc = make_predictions(matches_rolling, (predictors+new_cols))

In [40]:
new_acc

0.6029411764705882

In [41]:
pd.crosstab(index=new_combined["actual"], columns=new_combined["prediction"])

prediction    0   1
actual             
0           207  33
1           111  49

In [42]:
207/318

0.6509433962264151

In [43]:
49/82

0.5975609756097561

In [57]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
rf_preds = rf.predict(test[predictors+new_cols])
cm = confusion_matrix(test["target"], rf_preds)
print(cm)
print('Accuracy:', accuracy_score(test["target"], rf_preds))
print('Precision:', precision_recall_fscore_support(test["target"], rf_preds, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(test["target"], rf_preds, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(test["target"], rf_preds, average=None)[2][0])

[[207  33]
 [111  49]]
Accuracy: 0.64
Precision: 0.6509433962264151
Recall: 0.8625
F1 Score: 0.7419354838709677


In [45]:
new_combined = new_combined.merge(matches_rolling[["Date", "Team", "Opponent", "Result"]], left_index=True, right_index=True)

In [46]:
new_combined

actual  prediction       Date                     Team         Opponent  \
80         1           1 2023-10-08                  Arsenal  Manchester City   
81         0           1 2023-10-21                  Arsenal          Chelsea   
82         1           1 2023-10-28                  Arsenal    Sheffield Utd   
83         0           1 2023-11-04                  Arsenal    Newcastle Utd   
84         1           1 2023-11-11                  Arsenal          Burnley   
...      ...         ...        ...                      ...              ...   
1976       1           0 2024-02-04  Wolverhampton Wanderers          Chelsea   
1977       0           0 2024-02-10  Wolverhampton Wanderers        Brentford   
1978       1           0 2024-02-17  Wolverhampton Wanderers        Tottenham   
1979       1           0 2024-02-25  Wolverhampton Wanderers    Sheffield Utd   
1980       0           0 2024-03-02  Wolverhampton Wanderers    Newcastle Utd   

     Result  
80        W  
81        D  
82        W  
83        L  
84        W  
...     ...  
1976      W  
1977      L  
1978      W  
1979      W  
1980      L  

[400 rows x 6 columns]

In [47]:
merged = new_combined.merge(new_combined, left_on=["Date", "Team"], right_on=["Date", "Opponent"])

In [48]:
merged

actual_x  prediction_x       Date           Team_x       Opponent_x  \
0           1             1 2023-10-08          Arsenal  Manchester City   
1           0             1 2023-10-21          Arsenal          Chelsea   
2           1             1 2023-10-28          Arsenal    Sheffield Utd   
3           0             1 2023-11-04          Arsenal    Newcastle Utd   
4           1             1 2023-11-11          Arsenal          Burnley   
..        ...           ...        ...              ...              ...   
236         1             1 2024-02-10  Manchester City          Everton   
237         0             1 2024-02-17  Manchester City          Chelsea   
238         1             1 2024-02-20  Manchester City        Brentford   
239         1             1 2024-02-24  Manchester City      Bournemouth   
240         1             1 2024-03-03  Manchester City   Manchester Utd   

    Result_x  actual_y  prediction_y             Team_y       Opponent_y  \
0          W         0             1    Manchester City          Arsenal   
1          D         0             0            Chelsea          Arsenal   
2          W         0             0   Sheffield United          Arsenal   
3          L         1             1   Newcastle United          Arsenal   
4          W         0             0            Burnley          Arsenal   
..       ...       ...           ...                ...              ...   
236        W         0             0            Everton  Manchester City   
237        D         0             0            Chelsea  Manchester City   
238        W         0             0          Brentford  Manchester City   
239        W         0             0        Bournemouth  Manchester City   
240        W         0             0  Manchester United  Manchester City   

    Result_y  
0          L  
1          D  
2          L  
3          W  
4          L  
..       ...  
236        L  
237        D  
238        L  
239        L  
240        L  

[241 rows x 11 columns]

In [49]:
merged[(merged["prediction_x"] == 1) & (merged["prediction_y"] == 0)]["actual_x"].value_counts()

actual_x
1    40
0    13
Name: count, dtype: int64

In [50]:
40/53

0.7547169811320755

## XG_Boost

In [51]:
from xgboost import XGBClassifier
xg_classifier = XGBClassifier()
train = matches_rolling[matches_rolling["Date"] < '2023-10-01']
test = matches_rolling[matches_rolling["Date"] > '2023-10-01']
xg_classifier.fit(train[predictors+new_cols],train["target"])

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...)

In [52]:
new_preds = xg_classifier.predict(test[predictors+new_cols])

In [53]:
acc = accuracy_score(test["target"], new_preds)

In [54]:
acc

0.6225

In [55]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
cm = confusion_matrix(test["target"], new_preds)
print(cm)
print('Accuracy:', accuracy_score(test["target"], new_preds))
print('Precision:', precision_recall_fscore_support(test["target"], new_preds, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(test["target"], new_preds, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(test["target"], new_preds, average=None)[2][0])

[[187  53]
 [ 98  62]]
Accuracy: 0.6225
Precision: 0.656140350877193
Recall: 0.7791666666666667
F1 Score: 0.7123809523809523


## Kernel Support Vector Machine model

In [56]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(train[predictors+new_cols],train["target"])

SVC(random_state=0)

In [58]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
pred = classifier.predict(test[predictors+new_cols])
cm = confusion_matrix(test["target"], pred)
print(cm)
print('Accuracy:', accuracy_score(test["target"], pred))
print('Precision:', precision_recall_fscore_support(test["target"], pred, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(test["target"], pred, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(test["target"], pred, average=None)[2][0])

[[220  20]
 [134  26]]
Accuracy: 0.615
Precision: 0.6214689265536724
Recall: 0.9166666666666666
F1 Score: 0.7407407407407407


## Applying K-fold Cross Validation

In [59]:
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = train[predictors+new_cols], y = train["target"], cv = 10)
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))

Accuracy: 64.41 %
Standard Deviation: 5.54 %


## Applying Grid Search

In [60]:
from sklearn.model_selection import GridSearchCV
parameters = [{'C': [0.25, 0.5, 0.75, 1], 'kernel': ['linear']},
              {'C': [0.25, 0.5, 0.75, 1], 'kernel': ['rbf'], 'gamma': [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]}]
grid_search = GridSearchCV(estimator = classifier,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
grid_search.fit(train[predictors+new_cols], train["target"])
best_accuracy = grid_search.best_score_
best_parameters = grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 65.80 %
Best Parameters: {'C': 0.25, 'kernel': 'linear'}


## Naieve Bayes

In [61]:
from sklearn.naive_bayes import GaussianNB
nb_classifier = GaussianNB()
nb_classifier.fit(train[predictors+new_cols], train["target"])

GaussianNB()

In [62]:

from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
nb_pred = nb_classifier.predict(test[predictors+new_cols])
cm = confusion_matrix(test["target"], nb_pred)
print(cm)
print('Accuracy:', accuracy_score(test["target"], nb_pred))
print('Precision:', precision_recall_fscore_support(test["target"], nb_pred, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(test["target"], nb_pred, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(test["target"], nb_pred, average=None)[2][0])

[[174  66]
 [ 87  73]]
Accuracy: 0.6175
Precision: 0.6666666666666666
Recall: 0.725
F1 Score: 0.6946107784431138


## K-Nearest Neighbours

In [63]:
from sklearn.neighbors import KNeighborsClassifier
knn_classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
knn_classifier.fit(train[predictors+new_cols], train["target"])

KNeighborsClassifier()

In [64]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
knn_pred = nb_classifier.predict(test[predictors+new_cols])
cm = confusion_matrix(test["target"], knn_pred)
print(cm)
print('Accuracy:', accuracy_score(test["target"], knn_pred))
print('Precision:', precision_recall_fscore_support(test["target"], knn_pred, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(test["target"], knn_pred, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(test["target"], knn_pred, average=None)[2][0])

[[174  66]
 [ 87  73]]
Accuracy: 0.6175
Precision: 0.6666666666666666
Recall: 0.725
F1 Score: 0.6946107784431138


## Artificial Neural Network

In [65]:
import tensorflow as tf

In [66]:
ann = tf.keras.models.Sequential()

# Adding the input layer and the first hidden layer
ann.add(tf.keras.layers.Dense(units=10, activation='relu'))

# Adding the second hidden layer
ann.add(tf.keras.layers.Dense(units=10, activation='relu'))


# Adding the output layer
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

In [67]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [68]:
ann.fit(train[predictors+new_cols], train["target"], batch_size = 32, epochs = 100)

Epoch 1/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 1s 651us/step - accuracy: 0.5424 - loss: 0.7754 
Epoch 2/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - accuracy: 0.6469 - loss: 0.6481
Epoch 3/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - accuracy: 0.6526 - loss: 0.6379
Epoch 4/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 529us/step - accuracy: 0.6429 - loss: 0.6421
Epoch 5/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 712us/step - accuracy: 0.6474 - loss: 0.6457
Epoch 6/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 529us/step - accuracy: 0.6422 - loss: 0.6479
Epoch 7/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - accuracy: 0.6496 - loss: 0.6432
Epoch 8/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - accuracy: 0.6640 - loss: 0.6235
Epoch 9/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - accuracy: 0.6418 - loss: 0.6511
Epoch 10/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - accuracy: 0.6258 - loss: 0.6444
Epoch 11/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - accuracy: 0.6404 - loss: 0.6521
Epoch 12/100
50/50 ━━━━━━━━━━━━━━━━━━━━ 

In [69]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
ann_pred = ann.predict(test[predictors+new_cols])
ann_pred = (ann_pred > 0.5)
cm = confusion_matrix(test["target"], ann_pred)
print(cm)
print('Accuracy:', accuracy_score(test["target"], ann_pred))
print('Precision:', precision_recall_fscore_support(test["target"], ann_pred, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(test["target"], ann_pred, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(test["target"], ann_pred, average=None)[2][0])

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
[[192  48]
 [ 91  69]]
Accuracy: 0.6525
Precision: 0.6784452296819788
Recall: 0.8
F1 Score: 0.734225621414914


In [70]:
ann_pred = ann.predict(test[predictors+new_cols])

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 499us/step


In [71]:
import numpy as np
np.squeeze(ann_pred).shape

(400,)

In [72]:
test["target"].shape

(400,)

In [73]:
ann_combined = pd.DataFrame(dict(actual=test["target"], prediction = np.squeeze(ann_pred)))
ann_combined = ann_combined.merge(matches_rolling[["Date", "Team", "Opponent", "Result"]], left_index=True, right_index=True)

In [74]:
ann_merged = ann_combined.merge(ann_combined, left_on=["Date", "Team"], right_on=["Date", "Opponent"])
ann_merged[(ann_merged["prediction_x"] > 0.5) & (ann_merged["prediction_y"] < 0.5)]["actual_x"].value_counts()

actual_x
1    46
0    18
Name: count, dtype: int64

In [76]:
46/64

0.71875

## Adding Opponent Rolling Averages as Indicators

In [84]:
new_matches = matches_rolling.merge(matches_rolling, left_on=["Date", "Team"], right_on=["Date", "Opponent"])

In [85]:
new_matches

Date Time_x          Comp_x       Round_x Day_x Venue_x Result_x  \
0    2021-09-11  15:00  Premier League   Matchweek 4   Sat    Home        W   
1    2021-09-18  15:00  Premier League   Matchweek 5   Sat    Away        W   
2    2021-09-26  16:30  Premier League   Matchweek 6   Sun    Home        W   
3    2021-10-02  17:30  Premier League   Matchweek 7   Sat    Away        D   
4    2021-10-18  20:00  Premier League   Matchweek 8   Mon    Home        D   
...         ...    ...             ...           ...   ...     ...      ...   
1282 2022-04-30  15:00  Premier League  Matchweek 35   Sat    Home        L   
1283 2022-05-07  15:00  Premier League  Matchweek 36   Sat    Away        L   
1284 2022-05-11  19:45  Premier League  Matchweek 30   Wed    Home        D   
1285 2022-05-15  14:00  Premier League  Matchweek 37   Sun    Home        L   
1286 2022-05-22  16:00  Premier League  Matchweek 38   Sun    Away        L   

      GF_x  GA_x      Opponent_x  ...  PK_rolling_y  PKatt_rolling_y  \
0      1.0   0.0    Norwich City  ...      0.333333         0.333333   
1      1.0   0.0         Burnley  ...      0.000000         0.000000   
2      3.0   1.0       Tottenham  ...      0.000000         0.000000   
3      0.0   0.0        Brighton  ...      0.333333         0.333333   
4      2.0   2.0  Crystal Palace  ...      0.333333         0.333333   
...    ...   ...             ...  ...           ...              ...   
1282   1.0   2.0         Burnley  ...      0.000000         0.333333   
1283   0.0   1.0  Crystal Palace  ...      0.000000         0.000000   
1284   0.0   0.0         Everton  ...      0.000000         0.000000   
1285   1.0   5.0  Leicester City  ...      0.000000         0.000000   
1286   1.0   2.0         Chelsea  ...      0.333333         0.333333   

      Blocks_rolling_y  Int_rolling_y Tkl+Int_rolling_y Clr_rolling_y  \
0            10.333333      10.666667         29.000000     27.333333   
1            12.333333      12.666667         29.666667     22.000000   
2            17.333333       9.666667         28.000000     24.000000   
3             9.000000      12.666667         34.000000     19.333333   
4            16.333333       9.333333         28.333333     16.333333   
...                ...            ...               ...           ...   
1282         12.333333       9.333333         24.666667     23.666667   
1283         13.333333       9.000000         23.000000     17.666667   
1284         16.000000      10.666667         26.000000     28.666667   
1285         13.333333       5.000000         23.666667     10.000000   
1286          6.666667       9.000000         24.666667     11.333333   

     SoTA_rolling_y Saves_rolling_y  OppCrosses_rolling_y  \
0          4.333333        1.333333             18.666667   
1          5.333333        3.333333             15.333333   
2          5.000000        3.000000             19.000000   
3          2.333333        1.666667             17.000000   
4          6.333333        4.333333             15.666667   
...             ...             ...                   ...   
1282       4.333333        4.000000             20.666667   
1283       3.000000        2.333333              8.000000   
1284       5.666667        4.666667             23.000000   
1285       5.333333        3.666667              8.333333   
1286       1.666667        0.666667              7.666667   

      CrossesStp_rolling_y  
0                 0.666667  
1                 1.000000  
2                 1.000000  
3                 2.666667  
4                 0.333333  
...                    ...  
1282              1.333333  
1283              1.000000  
1284              1.333333  
1285              0.000000  
1286              0.333333  

[1287 rows x 117 columns]

In [86]:
new_predictors = [f"{c}_x" for c in predictors]

In [87]:
x_cols = [f"{c}_x" for c in new_cols]

In [89]:
y_cols = [f"{c}_y" for c in new_cols]

In [93]:
def make_new_predictions(data, predictors):
    train = data[data["Date"] < '2023-10-01']
    test = data[data["Date"] > '2023-10-01']
    rf.fit(train[predictors],train["target_x"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test["target_x"], prediction = preds))
    accuracy_score(test["target_x"], preds)
    return combined, acc

In [94]:
xy_combined, xy_acc = make_new_predictions(new_matches, (new_predictors+x_cols+y_cols))

In [95]:
xy_acc

0.6225

In [97]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
xy_test = new_matches[new_matches["Date"] > '2023-10-01']
xy_preds = rf.predict(xy_test[new_predictors+x_cols+y_cols])
cm = confusion_matrix(xy_test["target_x"], xy_preds)
print(cm)
print('Accuracy:', accuracy_score(xy_test["target_x"], xy_preds))
print('Precision:', precision_recall_fscore_support(xy_test["target_x"], xy_preds, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(xy_test["target_x"], xy_preds, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(xy_test["target_x"], xy_preds, average=None)[2][0])

[[128  11]
 [ 60  42]]
Accuracy: 0.7053941908713693
Precision: 0.6808510638297872
Recall: 0.920863309352518
F1 Score: 0.7828746177370031


In [98]:
128/188

0.6808510638297872

In [100]:
170/241

0.7053941908713693

In [101]:
xy_combined = xy_combined.merge(matches_rolling[["Date", "Team", "Opponent", "Result"]], left_index=True, right_index=True)

In [102]:
merged = xy_combined.merge(xy_combined, left_on=["Date", "Team"], right_on=["Date", "Opponent"])

In [103]:
merged[(merged["prediction_x"] == 1) & (merged["prediction_y"] == 0)]["actual_x"].value_counts()

actual_x
1    10
0     3
Name: count, dtype: int64

In [104]:
10/13

0.7692307692307693

## Retraining ANN with extra data

In [171]:
new_ann = tf.keras.models.Sequential()

# Adding the input layer and the first hidden layer
new_ann.add(tf.keras.layers.Dense(units=12, activation='relu'))

# Adding the second hidden layer
new_ann.add(tf.keras.layers.Dense(units=12, activation='relu'))

new_ann.add(tf.keras.layers.Dense(units=12, activation='relu'))

# Adding the output layer
new_ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

In [172]:
new_ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [173]:
train = new_matches[new_matches["Date"] < '2023-10-01']

In [174]:
new_ann.fit(train[new_predictors+x_cols+y_cols], train["target_x"], batch_size = 32, epochs = 100)

Epoch 1/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 623us/step - accuracy: 0.5937 - loss: 0.8522 
Epoch 2/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 561us/step - accuracy: 0.6742 - loss: 0.6432
Epoch 3/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 561us/step - accuracy: 0.6484 - loss: 0.6524
Epoch 4/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 592us/step - accuracy: 0.6427 - loss: 0.6327
Epoch 5/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 623us/step - accuracy: 0.6515 - loss: 0.6092
Epoch 6/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 592us/step - accuracy: 0.6655 - loss: 0.6124
Epoch 7/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 592us/step - accuracy: 0.6560 - loss: 0.6131
Epoch 8/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 561us/step - accuracy: 0.6883 - loss: 0.5896
Epoch 9/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 592us/step - accuracy: 0.6740 - loss: 0.6108
Epoch 10/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 592us/step - accuracy: 0.7014 - loss: 0.5925
Epoch 11/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 560us/step - accuracy: 0.6943 - loss: 0.5977
Epoch 12/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 

In [175]:
new_ann_pred = new_ann.predict(xy_test[new_predictors+x_cols+y_cols])

8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


In [176]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
new_ann_pred = (new_ann_pred > 0.5)
cm = confusion_matrix(xy_test["target_x"], new_ann_pred)
print(cm)
print('Accuracy:', accuracy_score(xy_test["target_x"], new_ann_pred))
print('Precision:', precision_recall_fscore_support(xy_test["target_x"], new_ann_pred, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(xy_test["target_x"], new_ann_pred, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(xy_test["target_x"], new_ann_pred, average=None)[2][0])

[[115  24]
 [ 53  49]]
Accuracy: 0.6804979253112033
Precision: 0.6845238095238095
Recall: 0.8273381294964028
F1 Score: 0.749185667752443


## Trying Recurrent Neural Network

In [131]:
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers import Dropout

In [133]:
regressor = Sequential()

# Adding the first LSTM layer and some Dropout regularisati
regressor.add(LSTM(units = 60, return_sequences = True, input_shape = (train.shape[1], 1)))
regressor.add(Dropout(0.2))

# Adding a second LSTM layer and some Dropout regularisation
regressor.add(LSTM(units = 60, return_sequences = True))
regressor.add(Dropout(0.2))

# Adding a third LSTM layer and some Dropout regularisation
regressor.add(LSTM(units = 60, return_sequences = True))
regressor.add(Dropout(0.2))

# Adding a fourth LSTM layer and some Dropout regularisation
regressor.add(LSTM(units = 60))
regressor.add(Dropout(0.2))

# Adding the output layer
regressor.add(Dense(units = 1))

C:\Users\lukeg\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\rnn\rnn.py:205: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [134]:
regressor.compile(optimizer = 'RMSprop', loss = 'mean_squared_error')


In [135]:
regressor.fit(train[new_predictors+x_cols+y_cols], train["target_x"], batch_size = 32, epochs = 100)

Epoch 1/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 4s 23ms/step - loss: 0.2877
Epoch 2/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2387
Epoch 3/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - loss: 0.2307
Epoch 4/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - loss: 0.2289
Epoch 5/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - loss: 0.2187
Epoch 6/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2265
Epoch 7/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2112
Epoch 8/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2093
Epoch 9/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2126
Epoch 10/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - loss: 0.2162
Epoch 11/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2167
Epoch 12/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - loss: 0.2187
Epoch 13/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2045
Epoch 14/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - loss: 0.2129
Epoch 15/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - lo

In [136]:
rnn_pred = regressor.predict(xy_test[new_predictors+x_cols+y_cols])

8/8 ━━━━━━━━━━━━━━━━━━━━ 1s 55ms/step


In [137]:
rnn_pred

array([[0.12242632],
       [0.31072995],
       [0.50031066],
       [0.45083046],
       [0.6342502 ],
       [0.66708183],
       [0.65202916],
       [0.5956141 ],
       [0.6382138 ],
       [0.24279419],
       [0.4596532 ],
       [0.4701224 ],
       [0.5658685 ],
       [0.5353726 ],
       [0.44102883],
       [0.43256664],
       [0.5582777 ],
       [0.6596182 ],
       [0.622416  ],
       [0.66492456],
       [0.4681961 ],
       [0.4983886 ],
       [0.58320135],
       [0.65648544],
       [0.69323444],
       [0.6754501 ],
       [0.33958018],
       [0.3008451 ],
       [0.12225254],
       [0.25260457],
       [0.4906168 ],
       [0.29168516],
       [0.63181657],
       [0.6782863 ],
       [0.6189899 ],
       [0.6575574 ],
       [0.64090276],
       [0.5793587 ],
       [0.535867  ],
       [0.51442456],
       [0.22071604],
       [0.33220792],
       [0.284463  ],
       [0.21449552],
       [0.28387305],
       [0.39555085],
       [0.27938807],
       [0.279

In [138]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
rnn_pred = (rnn_pred > 0.5)
cm = confusion_matrix(xy_test["target_x"], rnn_pred)
print(cm)
print('Accuracy:', accuracy_score(xy_test["target_x"], rnn_pred))
print('Precision:', precision_recall_fscore_support(xy_test["target_x"], rnn_pred, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(xy_test["target_x"], rnn_pred, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(xy_test["target_x"], rnn_pred, average=None)[2][0])

[[124  15]
 [ 64  38]]
Accuracy: 0.6721991701244814
Precision: 0.6595744680851063
Recall: 0.8920863309352518
F1 Score: 0.7584097859327217


## Trying ANN and RNN layers in same model

In [187]:
nn = Sequential()

# Adding the first LSTM layer and some Dropout regularisati
nn.add(LSTM(units = 60, return_sequences = True, input_shape = (train.shape[1], 1)))
nn.add(Dense(units=60, activation='relu'))
nn.add(Dropout(0.2))

# Adding a second LSTM layer and some Dropout regularisation
nn.add(LSTM(units = 60, return_sequences = True))
nn.add(Dense(units=60, activation='relu'))
nn.add(Dropout(0.2))

# Adding a third LSTM layer and some Dropout regularisation
nn.add(LSTM(units = 60, return_sequences = True))
nn.add(Dense(units=60, activation='relu'))
nn.add(Dropout(0.2))

# Adding a fourth LSTM layer and some Dropout regularisation
nn.add(LSTM(units = 60))
nn.add(Dense(units=60, activation='relu'))
nn.add(Dropout(0.2))

# Adding the output layer
nn.add(Dense(units = 1))

C:\Users\lukeg\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\rnn\rnn.py:205: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [188]:
nn.compile(optimizer = 'RMSprop', loss = 'mean_squared_error')


In [189]:
nn.fit(train[new_predictors+x_cols+y_cols], train["target_x"], batch_size = 32, epochs = 100)

Epoch 1/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 4s 24ms/step - loss: 0.2937
Epoch 2/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2404
Epoch 3/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2437
Epoch 4/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2251
Epoch 5/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2399
Epoch 6/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2446
Epoch 7/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 25ms/step - loss: 0.2366
Epoch 8/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2358
Epoch 9/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2369
Epoch 10/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2362
Epoch 11/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2268
Epoch 12/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2229
Epoch 13/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2177
Epoch 14/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - loss: 0.2144
Epoch 15/100
33/33 ━━━━━━━━━━━━━━━━━━━━ 1s 24ms/step - lo

In [190]:
nn_pred = nn.predict(xy_test[new_predictors+x_cols+y_cols])

8/8 ━━━━━━━━━━━━━━━━━━━━ 1s 56ms/step


In [191]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
nn_pred = (nn_pred > 0.5)
cm = confusion_matrix(xy_test["target_x"], nn_pred)
print(cm)
print('Accuracy:', accuracy_score(xy_test["target_x"], nn_pred))
print('Precision:', precision_recall_fscore_support(xy_test["target_x"], nn_pred, average=None)[0][0])
print('Recall:', precision_recall_fscore_support(xy_test["target_x"], nn_pred, average=None)[1][0])
print('F1 Score:', precision_recall_fscore_support(xy_test["target_x"], nn_pred, average=None)[2][0])

[[116  23]
 [ 53  49]]
Accuracy: 0.6846473029045643
Precision: 0.6863905325443787
Recall: 0.8345323741007195
F1 Score: 0.7532467532467533


In [193]:
nn_pred = nn.predict(xy_test[new_predictors+x_cols+y_cols])
nn_combined = pd.DataFrame(dict(actual=xy_test["target_x"], prediction = np.squeeze(nn_pred)))
nn_combined = nn_combined.merge(matches_rolling[["Date", "Team", "Opponent", "Result"]], left_index=True, right_index=True)

8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step


In [195]:
nn_merged = nn_combined.merge(nn_combined, left_on=["Date", "Team"], right_on=["Date", "Opponent"])
nn_merged[(nn_merged["prediction_x"] > 0.5) & (nn_merged["prediction_y"] < 0.5)]["actual_x"].value_counts()

actual_x
1    8
0    5
Name: count, dtype: int64

# Conclusion

While all models tried performed quite well given the complexity of the Premier League, Random Forest Walk seems to have the slightest of edges on the other models, particularly in how it is scaling with extra data, reaching an accuracy of 70% with precision of 68%. It also had a better success rate when combining the 2 predictions, and only considering a win prediction when the opposing team was also predicted to lose.
It is worth noting that various Deep Learning models also performed well with a marginally better precision but slightly worse accuracy compared to random forest walk. The addition of recurrent layers had a marginal effect on the ANN, but did improve accuracy and precision ever so slightly.